wagtail/models/__init__.py

```diff
@@ -177,7 +177,11 @@ def get_streamfield_names(model_class):
 
 class BasePageManager(models.Manager):
     def get_queryset(self):
-        return self._queryset_class(self.model).order_by("path")
+        qs = self._queryset_class(self.model).order_by("path")
+        # Restrict proxy model querysets to the exact type
+        if self.model._meta.proxy:
+            return qs.exact_type(self.model)
+        return qs
```

Comment:
Thanks for revisiting this feature! I tried out this PR with some simple models and everything seems to work as expected.

I spent some time trying to understand why this specific change to `get_queryset` was necessary. My mental model of proxy models has been "like a child model, but without a separate table", but this is not quite right (and unfortunately [the Django documentation on this topic](https://docs.djangoproject.com/en/4.1/topics/db/models/#proxy-models) doesn't quite state this as explicitly as it could). So for example given these models:

```py
class Parent(models.Model):
    name = models.TextField(blank=True, null=True)

class Child(Parent):
    pass

class ProxyChild(Parent):
    class Meta:
        proxy = True
```

The default Django behavior is surprising if you think of the proxy class as a "child":

```py
>>> from home.models import *
>>> Parent.objects.create(name="parent")
<Parent: Parent object (1)>
>>> Child.objects.create(name="child")
<Child: Child object (2)>
>>> ProxyChild.objects.create(name="proxy")
<ProxyChild: ProxyChild object (3)>
>>> Parent.objects.count()
3
# This makes sense if Child and ProxyChild are also Parents.
>>> Child.objects.count()
1
# This makes sense, I only created one Child object.
>>> ProxyChild.objects.count()
3
# Wait, what?
```

But this does make sense if you understand that a "proxy" model is just another view into the same table, not really a "child" of a parent.

Back to the specific code change to `get_queryset` here, although it is required to make proxy Page models work as expected in Wagtail, does seem to make things inconsistent with that basic Django behavior. So creating the same page types like above:

```py
class ParentPage(Page):
    name = models.TextField(blank=True, null=True)

class ChildPage(ParentPage):
    pass

class ProxyChildPage(ParentPage):
    class Meta:
        proxy = True
```

Now the logic works differently:

```py
>>> root.add_child(instance=ParentPage(title="parent", slug="parent"))
<ParentPage: parent>
>>> root.add_child(instance=ChildPage(title="child", slug="child"))
<ChildPage: child>
>>> root.add_child(instance=ProxyChildPage(title="proxy", slug="proxy"))
<ProxyChildPage: proxy>
>>> ParentPage.objects.count()
3
>>> ChildPage.objects.count()
1
>>> ProxyChildPage.objects.count()
1
# This behaves differently than above; now proxy models really are children.
```

Maybe this is okay, since it's more consistent with how users might want to use Page proxy models in Wagtail? But then, what implications does this have for non-Page proxy models in Wagtail?

It seems that adding the logic here in the base manager `get_queryset` _should_ be sufficient to guard all downstream access to the proxy queryset... but is it possible there are some other paths in Django code where the queryset gets constructed in some other way?

Reply:
any inherited classes as proxy, works exactly as the inherited class

as it is considered a shadow for the main class(inherited one)

Reply:
> Maybe this is okay, since it's more consistent with how users might want to use Page proxy models in Wagtail? But then, what implications does this have for non-Page proxy models in Wagtail?

Pages are a bit of a special case, because: 

1. We are in control of the `Page` model
2. We are already customising the `PageManager` and providing very custom query-related functionality via `PageQuerySet` 
3. We already store the `ContentType` for every page, so so can easily tell the difference between  a `ParentPage`, a `ChildPage` and a `ProxyChildPage`. regardless of what type of inheritance is being used
4. We are already using the `content_type` field value, the `Page.specific` property, and the `PageQuerySet.specific()` method to change interfaces and behaviour depending on the specific page type.

This approach (or one similar) is often actually taken by developers who want to make use of proxy models too. For example, [see this popular article](https://www.benlopatin.com/using-django-proxy-models), where custom managers are added to return the relevant types of items based on a `type` `CharField` value. 

We can't give snippets and other models the same treatment, because we typically have no control over the models used, and so there is no (nice) way to hack the manager/queryset behaviour. For developers who want to use proxy models, it's really up to them to understand how they work, and what they want to do about this sort of thing. We can provide guidance in the docs to help manage expectations, but that's where our involvement should end IMO.